juju/juju
# Status history

The status history is now a dependency for the service factory. This
allows us to set history at the service level. The status history
worker uses the logsink worker as the backing store. By using an
intermediary we can change the backend to the status history without
updating the dependencies of each.

The roadblock to this change was that the model logger required
the model uuid, model owner and model name. There was a latent bug
where if you changed the model owner and model name, but used the
same model uuid it would give you a cached model logger to file you
might not want.

The logsink also can't depend on the service factory because we want
to use the logsink in the service factory. To break the cycle, I
pushed the controller config values into the agent config. This is
a common pattern that we've used before. The prior logsink worker
didn't update the logger size and rotation information if it changed,
only if the agent was rebooted. I've mirrored what we currently
have, although it's relatively simple to update the agent config
updater to include those values, which would do it automatically.

## Checklist




## QA steps

```sh
$ juju bootstrap lxd test
$ juju add-model default
$ juju deploy ubuntu
```

We should see the statushistory log into the model log.

```sh
$ juju ssh -m controller 0
> cat /var/log/juju/models/<model path>.log
```

## Links

**Jira card:** [JUJU-5888](https://warthogs.atlassian.net/browse/JUJU-5888)



[JUJU-5888]: https://warthogs.atlassian.net/browse/JUJU-5888?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
